app/emoji.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"bytes"
+	"image"
+	"image/color/palette"
+	"image/draw"
+	"image/gif"
+	_ "image/jpeg"
+	"image/png"
+	"io"
+	"mime/multipart"
+	"net/http"
+
+	"github.com/disintegration/imaging"
+	"github.com/mattermost/platform/model"
+)
+
+const (
+	MaxEmojiFileSize = 1000 * 1024 // 1 MB
+	MaxEmojiWidth    = 128
+	MaxEmojiHeight   = 128
+)
+
+func CreateEmoji(sessionUserId string, emoji *model.Emoji, multiPartImageData *multipart.Form) (*model.Emoji, *model.AppError) {
+	// wipe the emoji id so that existing emojis can't get overwritten
+	emoji.Id = ""
+
+	// do our best to validate the emoji before committing anything to the DB so that we don't have to clean up
+	// orphaned files left over when validation fails later on
+	emoji.PreSave()
+	if err := emoji.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if emoji.CreatorId != sessionUserId {
+		return nil, model.NewAppError("createEmoji", "api.emoji.create.other_user.app_error", nil, "", http.StatusUnauthorized)
+	}
+
+	if result := <-Srv.Store.Emoji().GetByName(emoji.Name); result.Err == nil && result.Data != nil {
+		return nil, model.NewAppError("createEmoji", "api.emoji.create.duplicate.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if imageData := multiPartImageData.File["image"]; len(imageData) == 0 {
+		err := model.NewLocAppError("Context", "api.context.invalid_body_param.app_error", map[string]interface{}{"Name": "createEmoji"}, "")
+		err.StatusCode = http.StatusBadRequest
+		return nil, err
+	} else if err := UploadEmojiImage(emoji.Id, imageData[0]); err != nil {
+		return nil, err
+	}
+
+	if result := <-Srv.Store.Emoji().Save(emoji); result.Err != nil {
+		return nil, result.Err
+	} else {
+		return result.Data.(*model.Emoji), nil
+	}
+}
+
+func UploadEmojiImage(id string, imageData *multipart.FileHeader) *model.AppError {
+	file, err := imageData.Open()
+	if err != nil {
+		return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.open.app_error", nil, "")
```

Comment:
Can we use `NewAppError` here with a StatusBadRequest code? Same with all the below `NewLocAppError` uses?